src/main/java/org/truffleruby/core/cast/DurationToMillisecondsNode.java

```diff
@@ -65,12 +64,24 @@ protected long durationNil(Nil duration) {
     }
 
     @Specialization
-    protected long duration(RubyDynamicObject duration) {
-        if (floatCastNode == null) {
-            CompilerDirectives.transferToInterpreterAndInvalidate();
-            floatCastNode = insert(NumericToFloatNodeGen.create());
+    protected long duration(RubyDynamicObject duration,
+            @Cached DispatchNode divmodNode,
+            @Cached DispatchNode sumNode,
+            @Cached ToLongNode toLongNode) {
+        try {
+            Object result = divmodNode.call(duration, "divmod", 1);
+            Object out = sumNode.call(result, "sum");
+            return toLongNode.execute(out) * 1000;
+        } catch (RaiseException e) {
+            if (e.getException().getMetaClass().toString().equals("NoMethodError")) {
```

Comment:
This could be `e.getException() instanceof RubyNoMethodError`, but with `check_funcall` we don't need this `catch` at all.